### PR TITLE
Define features & first implementation

### PR DESCRIPTION
**Pull request by stefandoorn**
Added cart / product page checks for reserved stock.

### CODE REVIEW

**Review by loevgaard — DISMISSED** [review]

**Comment by stefandoorn** [comment]
@loevgaard Could you review again?

File: src/Repository/InCartQuantityForProductVariantOrderItemRepositoryAwareInterface.php [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+declare(strict_types=1);
+
+namespace Setono\SyliusReserveStockPlugin\Repository;
+
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Order\Model\OrderInterface;
+
+interface InCartQuantityForProductVariantOrderItemRepositoryAwareInterface
+{
+    public function inCartQuantityForProductVariantExcludingOrder(ProductVariantInterface $productVariant, ?OrderInterface $order): int;
```
Comment (loevgaard) [inline src/Repository/InCartQuantityForProductVariantOrderItemRepositoryAwareInterface.php:21]: Add phpdoc explaining this method

File: src/Repository/ProductVariantCartOrderItem.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace Setono\SyliusReserveStockPlugin\Repository;
+
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Order\Model\OrderInterface;
+
+trait ProductVariantCartOrderItem
+{
+    public function inCartQuantityForProductVariantExcludingOrder(
```
Comment (loevgaard) [inline src/Repository/ProductVariantCartOrderItem.php:16]: add phpdoc

File: src/Repository/ProductVariantCartOrderItem.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        if (null !== $order && null !== $order->getId()) {
+            $qb = $qb
+                ->andWhere('o.order != :order')
+                ->setParameter('order', $order);
+        }
+
+        return $this->produceSingleScalarQueryIntegerResult($qb);
+    }
+
+    private function createOrderItemVariantQueryBuilder(ProductVariantInterface $productVariant): QueryBuilder
```
Comment (loevgaard) [inline src/Repository/ProductVariantCartOrderItem.php:35]: Add phpdoc

File: src/Repository/ProductVariantCartOrderItem.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        return $qb
+            ->select('SUM(o.quantity) AS quantity')
+            ->innerJoin('o.order', 'cart')
+            ->andWhere('cart.state = :state')
+            ->andWhere('o.variant = :variant')
+            ->setParameter('state', OrderInterface::STATE_CART)
+            ->setParameter('variant', $productVariant);
+    }
+
+    private function produceSingleScalarQueryIntegerResult(QueryBuilder $queryBuilder): int
```
Comment (loevgaard) [inline src/Repository/ProductVariantCartOrderItem.php:54]: Add phpdoc

File: src/Repository/Stubs/ProductVariantCartOrderItemStub.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+declare(strict_types=1);
+
+namespace Setono\SyliusReserveStockPlugin\Repository\Stubs;
+
+use Setono\SyliusReserveStockPlugin\Repository\InCartQuantityForProductVariantOrderItemRepositoryAwareInterface;
+use Setono\SyliusReserveStockPlugin\Repository\ProductVariantCartOrderItem;
+use Sylius\Bundle\OrderBundle\Doctrine\ORM\OrderItemRepository;
+
+final class ProductVariantCartOrderItemStub extends OrderItemRepository implements InCartQuantityForProductVariantOrderItemRepositoryAwareInterface
```
Comment (loevgaard) [inline src/Repository/Stubs/ProductVariantCartOrderItemStub.php]: Should this be included in the plugin or can it be done any other way? It seems it's only used in the specs and not in the actual plugin

File: tests/Application/app/config/services.yml [hunk 6]
```diff
@@ -0,0 +1 @@
+services:
```
Comment (loevgaard) [inline tests/Application/app/config/services.yml]: I suppose you have an idea with this? :)

File: tests/Application/src/Repository/OrderItemRepository.php [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+declare(strict_types=1);
+
+namespace TestApp\Setono\SyliusReserveStockPlugin\Repository;
```
Comment (loevgaard) [inline tests/Application/src/Repository/OrderItemRepository.php]: I am not sure why this new namespace was introduced?

**Review by loevgaard — DISMISSED** [review]

**Review by loevgaard — APPROVED** [review]

**Pull request merged**